Convert SAM SoC GPIO drivers to DT_INST

nandojve opened the pull request:
This convert Atmel SAM/SAM0 GPIO drivers to use DT_INST defines.

Uses commit from: #23335

mbolivar-nordic committed:
scripts: gen_defines: augment edtlib nodes

Add additional attributes to each edtlib.Node we process, before
calling into the write_foo() routines.

This includes the identifier returned by node_ident(), which is used
as the primary identifier for the node, as well as lists for instance
and aliases nodes, and a catchall list that contains all the other
identifiers in addition to the primary one.

Use this information in a new for_each_ident(node, ident) def that
will be put to use in the various write_foo() routines.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
scripts: gen_defines: re-work write_regs() with augmented nodes

Group the macros together by namespace rather than putting all the
BASE_ADDRESS macros together and all the SIZE macros together. E.g.,
all the DT_INST_<x> namespace macros for each node now appear
consecutively.

Add a comment making it clear that this output comes from "regs",
since "BASE_ADDRESS" and "SIZE" are not property names.

Other than the order in which they appear and comments, the output
before and after this patch should be exactly the same.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
scripts: gen_defines: re-work write_irqs() with augmented nodes

Use augmented nodes to print macros grouped by namespace.

Other than the order in which they appear and comments, the output
before and after this patch should be exactly the same.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
scripts: gen_defines: re-work write_clocks() with augmented nodes

Mirror the change already done to write_regs().

Other than the order in which they appear and comments, the output
before and after this patch should be exactly the same.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

galak committed:
drivers: wifi: eswifi: Convert to use DT_INST defines

Use DT_INST define as this is the preferred set of defines to utilize
for drivers.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

mbolivar-nordic committed:
scripts: gen_defines: re-work write_props() with augmented nodes

This is similar to the work already done for regs.

Other than the order in which they appear and comments, the output
before and after this patch should be exactly the same.

We're intentionally leaving some of the helpers in module scope here
to use some of the subroutines elsewhere later on when reworking
write_spi_dev().

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
scripts: gen_defines: re-work write_spi_dev() with augmented nodes

This uses augmented nodes in the same way as done previously.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
scripts: gen_defines: re-work write_bus() with augmented nodes

This too is an attempt to reimplement the previous behavior exactly,
modulo the order in which things are defined.

This is the last function which is calling into the previous
implementation's out_node and node_*_alias() functions, so these can
be removed now.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
scripts: gen_defines: remove out()'s aliases kwarg

Write the same results explicitly in its only remaining caller.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
edtlib: add 'labels' attribute to Node class

This corresponds to the attribute by the same name in dtlib.Node.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
scripts: gen_defines: tweak out_define() whitespace

Pad node identifiers to 60 characters. This results in better
alignment in practice than the current value of 40, which is a bit
low.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

galak committed:
drivers: clock_control: rv32m1_pcc: convert to DT_INST defines

Convert driver to use DT_INST_ defines.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: clock_control: mcux_ccm: convert to DT_INST defines

Convert driver to use DT_INST_ defines.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: clock_control: mcux_mcg: convert to DT_INST defines

Convert driver to use DT_INST_ defines.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: clock_control: mcux_pcc: convert to DT_INST defines

Convert driver to use DT_INST_ defines.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: clock_control: mcux_sim: convert to DT_INST defines

Convert driver to use DT_INST_ defines.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: watchdog: sam: convert to DT_INST defines

Convert driver to use DT_INST_ defines.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: modem: ublox-sara-r4: convert to DT_INST defines

Convert driver to use DT_INST_ defines.  The preferred defines for
drivers are DT_INST_.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: dma: dma_dw: convert to DT_INST defines

Convert driver to use DT_INST_ defines.  The preferred defines for
drivers are DT_INST_.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: entropy: mcux_trng: convert to DT_INST defines

Convert driver to use DT_INST_ defines.  The preferred defines for
drivers are DT_INST_.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: entropy: sam: convert to DT_INST defines

Convert driver to use DT_INST_ defines.  The preferred defines for
drivers are DT_INST_.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: watchdog: wdt_mcux: convert to DT_INST defines

Convert driver to use DT_INST_ defines.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: mcux_gpt: convert to DT_INST defines

Convert driver to use DT_INST_ defines and remove Kconfig per instance
enablement in favor of DT_INST_ define existing.  Also, remove the
aliases that had been used for this driver in nxp_rt.dtsi.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: mcux_rtc: convert to DT_INST defines

Convert driver to use DT_INST_ defines.  Replace dts_fixup.h use
for DT_RTC_0_NAME with DT_INST_0_NXP_KINETIS_RTC_LABEL to be
consistent.  Also, remove the aliases that had been used for this
driver in various nxp_k*.dtsi.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: usb: usb_dw: convert to DT_INST defines

Convert driver to use DT_INST_ defines.  The preferred defines for
drivers are DT_INST_.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
driver: gpio: dw: convert to DT_INST defines

Convert driver to use DT_INST_ defines.  The preferred defines for
drivers are DT_INST_.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
board: hsdk: Convert defines into dts for display

The ILITEK ILI9340 should have been in the dts and added as #defines in
dts_fixup.h.  Fix this by adding a display node in the dts.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
driver: spi: dw: convert to DT_INST defines

Convert driver to use DT_INST_ defines.  The preferred defines for
drivers are DT_INST_.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
driver: interrupt_controller: dw: convert to DT_INST defines

Convert driver to use DT_INST_ defines.  The preferred defines for
drivers are DT_INST_.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

ulfalizer committed:
dts: edtlib/gen_defines: Fix API design re. dtc flags

It's better to allow per-instance EDT configuration than to set a global
variable on the edtlib module. Enable/disable the warning for reg/unit
address mismatches via a flag to EDT.__init__(), instead of via a global
variable. That makes it consistent too.

Another option would be to pass the 'dtc' flags to EDT.__init__(), but
it makes the interface a bit ugly. Maybe if it needs to emulate lots of
other flags later.

Clarify that edtlib itself isn't meant to have any state in the comment
at the top of the module.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

galak committed:
kconfig: Add dt_nodelabel_enabled function

Add a function takes a 'label' and returns "y" if we find an
"enabled" node that has a 'nodelabel' of 'label' in the EDT
otherwise we return "n"

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: counter: mchp_xec: convert to DT_INST defines

Convert driver to use DT_INST_ defines.  The preferred defines for
drivers are DT_INST_.

As part of this change we utilize the device tree for GIRQ info and
rename timer3 to 2 since we are doing this by instance number.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
dts: microchip,xec-timer: Add properties for GIRQ info

Add girq and girq-bit to encode per device information.  This allows the
driver to get any device unique info from device tree.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: counter: mchp_xec: Fix dts prop type for max-value

The max-value should just be an int and not an array.  Change the type
to 'int' in the binding and fixup the driver to match.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: espi: mchp_xec: convert to DT_INST defines

Convert driver to use DT_INST_ defines.  The preferred defines for
drivers are DT_INST_.  The driver mostly used DT_INST_ defines but
a few IRQ priority defines needed conversion.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
dts: microchip,xec-ps2: Add properties for GIRQ info

Add girq and girq-bit to encode per device information.  This allows the
driver to get any device unique info from device tree.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: ps2: mchp_xec: convert to DT_INST defines

Convert driver to use DT_INST_ defines.  The preferred defines for
drivers are DT_INST_.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: i2c: nios2: convert to DT_INST defines

Convert driver to use DT_INST_ defines.  The preferred defines for
drivers are DT_INST_.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

ulfalizer committed:
dts: edtlib: Turn global spi_dev_cs_gpio() into Node.spi_cs_gpio

spi_dev_cs_gpio() takes a Node and returns the chip select GPIO for it.
Having that information available directly from Node is neater, so turn
it into a Node.spi_cs_gpio property instead.

That gets rid of the only public global function in edtlib, which might
make the API design clearer too.

Tested with the sensortile_box board, which uses SPI chip select.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

galak committed:
drivers: ieee802154: mcr20a: convert to DT_INST defines

Convert driver to use DT_INST_ defines.  There was just one case for
CS_GPIOS that wasn't using DT_INST defines already.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: uart: esp32: convert to DT_INST defines

Convert driver to fully use DT_INST_ defines.  Currently IRQs are not
generated from DTS on esp32 so we create #defines for what they should
look like.  Convert the IRQ defines to match DT_INST style so if/when
we have that info in DTS it will look the same.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: uart: msp432p4xx: convert to DT_INST defines

Convert driver to fully use DT_INST_ defines.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: uart: native_posix: convert to DT_INST defines

Convert driver to fully use DT_INST_ defines.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: intc: plic: convert to DT_INST defines

Convert driver to use DT_INST_ defines.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

nandojve committed:
drivers: gpio: sam0: convert to DT_INST defines

Convert driver to use DT_INST_ defines.

see #23107

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
drivers: gpio: sam: convert to DT_INST defines

Convert driver to use DT_INST_ defines.

see #23107

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

mnkp commented on drivers/gpio/gpio_sam0.c:
```diff
@@ -299,86 +299,64 @@ static const struct gpio_driver_api gpio_sam0_api = {
+			DT_INST_##n##_ATMEL_SAM0_GPIO_NGPIOS),		  \
+	},								  \
+	.regs = (PortGroup *) DT_INST_##n##_ATMEL_SAM0_GPIO_BASE_ADDRESS, \
+	.id = n,							  \
```
Looking at the implementation of `sam0_eic_acquire` -> `sam0_eic_map_to_line` function it seems like the assignment of the id to the gpio instance is strict, i.e. id 0 corresponds to Port A.

nandojve replied:
This was not the first case. I've been solving following #23335 proposal and then with b36c80c it's possible solve that kind of issue.

mnkp commented on drivers/gpio/gpio_sam0.c:
```diff
@@ -299,86 +299,64 @@ static const struct gpio_driver_api gpio_sam0_api = {
-		.port_pin_mask = GPIO_PORT_PIN_MASK_FROM_NGPIOS(DT_INST_0_ATMEL_SAM0_GPIO_NGPIOS),
-	},
-	.regs = (PortGroup *)DT_ATMEL_SAM0_GPIO_PORT_A_BASE_ADDRESS,
 #ifdef CONFIG_SAM0_EIC
```
To avoid defining the `GPIO_SAM0_DEVICE_CONFIG` macro twice we could do 
```
#ifdef CONFIG_SAM0_EIC
#define GPIO_SAM0_EIC_ID .id = n,
#else
#define GPIO_SAM0_EIC_ID
#endif
```
and then use `GPIO_SAM0_EIC_ID` in `GPIO_SAM0_DEVICE_CONFIG`.

nandojve replied:
Done!

nandojve committed:
dts: bindings: base.yaml: Add hw instance id

Durant the process to convert Atmel drivers to use DT_INST_ defines
was detected the necessity of a hardware instance identificator.
Some drivers need configure port pins, clocks, DMA etc. This will
track the hw instance on the DT_INST_x_ independent of the 'x' value.

This will keep simple the conversion this add a new non mandatory
property called <id>. The driver that need these feature need add
'required=true' on the dts/binding/driver  and complement any
device tree with proper value.

Examples:

foo0: foo@1000 {
	compatible = "man,foo-drv";
	id = <0>;
	status="disabled";
};
foo1: foo@1001 {
	compatible = "man,foo-drv";
	id = <1>;
};
barn: bar@2001 {
	compatible = "man,bar-drv";
	id = <10>;
};
barm: bar@3001 {
	compatible = "man,bar-drv";
	id = <20>;
};

Will generate:

DT_INST_0_MAN_FOO_DRV_ID=1
DT_INST_0_MAN_BAR_DRV_ID=10
DT_INST_1_MAN_BAR_DRV_ID=20

It is intent to be used as below to propagate the hardware id on all
driver defines.

DRV_FOO_DEVICE_INIT(0, DT_INST_0_MAN_FOO_DRV_ID);

A good driver example that need this definition:

static const struct drv_foo_dev_cfg drv##n##_sam_config = {	\
	.regs = (Drv *)DT_INST_##n##_MAN_DRV_FOO_BASE_ADDRESS,	\
	.pin_rx = PIN_FOO##m##_RXD,				\
	.pin_tx = PIN_FOO##m##_TXD,				\
}

If 'm' isn't defined the driver will invariably configure the wrong
pins and communication won't happen. This can't be derived by node
name for many reasons. For instance. the above DT snip defined 'barn'
and 'barm' that have a name without number and an unusual value. It
represents two instances of bar that shouldn't assume values other
than defined on the id property.

see #23107

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
dts: bindings: gpio: atmel,sam0-gpio: Add hw id prop

Add DT HW ID for Atmel SAM0 GPIO driver.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve commented on drivers/gpio/gpio_sam.c:
```diff
@@ -321,142 +321,69 @@ int gpio_sam_init(struct device *dev)
-	.periph_id = DT_GPIO_SAM_PORTE_PERIPHERAL_ID,
-	.config_func = port_e_sam_config_func,
-};
+#if DT_INST_0_ATMEL_SAM_GPIO_BASE_ADDRESS
```
Done!

nandojve commented on drivers/gpio/gpio_sam.c:
```diff
@@ -321,142 +321,69 @@ int gpio_sam_init(struct device *dev)
+GPIO_SAM_DEVICE_CONFIG(0);
+GPIO_SAM_DEVICE_DATA(0);
+GPIO_SAM_DEVICE_INIT(0);
+GPIO_SAM_DEVICE_FUNC(0);
```
Am I missing something or will checkpath always complain with this type of construction?

mnkp commented on drivers/gpio/gpio_sam.c:
```diff
@@ -321,142 +321,62 @@ int gpio_sam_init(struct device *dev)
+			    &port_##n##_sam_config, POST_KERNEL,	\
+			    CONFIG_KERNEL_INIT_PRIORITY_DEVICE,		\
+			    &gpio_sam_api);				\
+	GPIO_SAM_IRQ_HANDLER(n)
```
If we write the macro as
```
#define GPIO_SAM_DEVICE_INIT(n)						\
static void port_##n##_sam_config_func(struct device *dev);		\
									\
static const struct gpio_sam_config port_##n##_sam_config = {		\
	.common = {							\
		.port_pin_mask = GPIO_PORT_PIN_MASK_FROM_NGPIOS(	\
			DT_INST_##n##_ATMEL_SAM_GPIO_NGPIOS),		\
	},								\
	.regs = (Pio *)DT_INST_##n##_ATMEL_SAM_GPIO_BASE_ADDRESS,	\
	.periph_id = DT_INST_##n##_ATMEL_SAM_GPIO_PERIPHERAL_ID,	\
	.config_func = port_##n##_sam_config_func,			\
};									\
									\
static struct gpio_sam_runtime port_##n##_sam_runtime;			\
									\
DEVICE_AND_API_INIT(port_##n##_sam,					\
		    DT_INST_##n##_ATMEL_SAM_GPIO_LABEL,			\
		    gpio_sam_init, &port_##n##_sam_runtime,		\
		    &port_##n##_sam_config, POST_KERNEL,		\
		    CONFIG_KERNEL_INIT_PRIORITY_DEVICE,			\
		    &gpio_sam_api);					\
									\
static void port_##n##_sam_config_func(struct device *dev)		\
{									\
	IRQ_CONNECT(DT_INST_##n##_ATMEL_SAM_GPIO_IRQ_0,			\
		    DT_INST_##n##_ATMEL_SAM_GPIO_IRQ_0_PRIORITY,	\
		    gpio_sam_isr, DEVICE_GET(port_##n##_sam), 0);	\
	irq_enable(DT_INST_##n##_ATMEL_SAM_GPIO_IRQ_0);			\
}
```
checkpatch will not complain.

mnkp commented on drivers/gpio/gpio_sam0.c:
```diff
@@ -299,86 +299,51 @@ static const struct gpio_driver_api gpio_sam0_api = {
-	.id = 0,
+#define GPIO_SAM0_EIC_ID(m) .id = m,
+#else
+#define GPIO_SAM0_EIC_ID(m)
```
```
#define GPIO_SAM0_EIC_ID(n) .id = DT_INST_##n##_ATMEL_SAM0_GPIO_ID,
#else
#define GPIO_SAM0_EIC_ID(n)
```

nandojve commented:
topic-devicetree already merged.
